### DISCUSSION TITLE
Can I hash a container?

### DISCUSSION THREAD

**Discussion by ianthehenry** [arximboldi/immer#249] [Q&A] [answered]
Is there a way to hash a container?

I can see that this listed as "future work" in https://github.com/arximboldi/immer/pull/19, and I wanted to know if that was done/how to do it. I can see that `std::hash` is still not defined, but with my extremely limited knowledge of C++ I cannot find an alternative way to hash containers.

Failing that, is there some way to stably iterate over the elements in a container such that I can construct my own law-abiding hash value? I don't care too badly about performance, but allocating a vector or something just to sort it feels wrong.

**Comment by arximboldi** [comment, marked as answer]
You can define override `std::hash` and define it by iterating over the container, hashing every element and combining the hashes (using something like `boost:::hash_combine` or your own hash combination formula).

**Reply by ianthehenry** [reply]
I had sort of assumed that this wouldn't work for unordered types, because the iteration order wouldn't be well-defined, and a standard hash mixer would be sensitive to that. But experimentally, playing with `immer::set`, it seems like the iteration order actually *is* consistent, regardless of how I construct the set. Is that guaranteed to be the case, or am I just getting lucky?

**Reply by arximboldi** [reply]
That is guaranteed to be the case. The order of iteration depends only on the hashes of the elements themselves, not the order of insertion/removal.
